tests/test_type_inference/test_attribute.py

```diff
@@ -106,3 +107,137 @@ def test_attribute_self_bind():
+        class A:
+            def __init__(self):
+                self.name = 'abc'
+
+        lst = [A(), A()]
+        x = lst[0].name
+        '''
+    module, ti = cs._parse_text(program, reset=True)
+    for assgn_node in module.nodes_of_class(astroid.AssignName):
+        if assgn_node.name == 'x':
+            x = ti.lookup_typevar(assgn_node, assgn_node.name)
+            assert ti.type_constraints.resolve(x).getValue() == str
+
+
+def test_unknown_class_attribute():
+    program = \
+        '''
+        def foo(a, x):
+            y = x.name
+            y = 3
+            z = x.name
+        '''
+    module, ti = cs._parse_text(program, reset=True)
+    for assgn_node in module.nodes_of_class(astroid.AssignName):
+        if assgn_node.name == 'z':
+            z = ti.lookup_typevar(assgn_node, assgn_node.name)
+            assert ti.type_constraints.resolve(z).getValue() == int
+
+
+def test_unknown_class_attribute2():
+    program = \
+        '''
+        def foo(x):
+            return x.name + 2
+            
+        def bar(y):
+            z = foo(y)
+        '''
+    module, ti = cs._parse_text(program, reset=True)
+    for binop_node in module.nodes_of_class(astroid.BinOp):
+        assert binop_node.inf_type.getValue() == int
+    for assgn_node in module.nodes_of_class(astroid.AssignName):
+        if assgn_node.name == 'z':
+            z = ti.lookup_typevar(assgn_node, assgn_node.name)
+            assert ti.type_constraints.resolve(z).getValue() == int
+
+
+def test_unknown_class_subscript_attribute():
+    raise SkipTest("Currently, the inferred type of the z variable is str")
+    program = \
+        '''
+        def foo(x):
+            return x.name[0]
+            
+        def bar(y):
+            z = foo(y)
+        '''
+    module, ti = cs._parse_text(program, reset=True)
+    for assgn_node in module.nodes_of_class(astroid.AssignName):
+        if assgn_node.name == 'z':
+            z = ti.lookup_typevar(assgn_node, assgn_node.name)
+            assert isinstance(ti.type_constraints.resolve(z).getValue(), TypeVar)
+
+
+def test_unknown_class_subscript_attribute2():
+    program = \
+        '''
+        def foo(x):
+            x.name[0] = 2
+            return x.name[1]
+
+        def bar(y):
+            z = foo(y)
+        '''
+    module, ti = cs._parse_text(program, reset=True)
+    for assgn_node in module.nodes_of_class(astroid.AssignName):
+        if assgn_node.name == 'z':
+            z = ti.lookup_typevar(assgn_node, assgn_node.name)
+            assert ti.type_constraints.resolve(z).getValue() == int
+
+
+def test_invalid_builtin_attribute():
+    program = \
+        '''
+        x = 3
+        y = x.name
+        '''
+    module, ti = cs._parse_text(program, reset=True)
+    assgn_node = list(module.nodes_of_class(astroid.Assign))[1]
+    assert isinstance(assgn_node.inf_type, TypeFail)
+
+
+def test_attribute_unification_fail():
+    raise SkipTest('This case is not supported yet')
```

Comment:
This case requires that when we unify a `TypeVar` with a class instance type or class type, we unify each common attribute for both of the types (although that won't be enough to make this pass due to the fact that the type variable in `unify_call` will be a fresh one). This will be a separate pull request.